ios/RN/RNCamera.m

```diff
@@ -421,40 +403,110 @@ - (void)record:(NSDictionary *)options resolve:(RCTPromiseResolveBlock)resolve r
-          self.videoCodecType = videoCodecType;
-        } else {
-          RCTLogWarn(@"%s: Video Codec '%@' is not supported on this device.", __func__, videoCodecType);
-        }
-      } else {
-        RCTLogWarn(@"%s: Setting videoCodec is only supported above iOS version 10.", __func__);
-      }
-    }
+    // Set flag that notifies 'didOutputSampleBuffer' delegate method to initialize writer with buffer presentation timestamp
+    self.canAppendBuffer = YES;
+    self.videoRecordedResolve = resolve;
+    self.videoRecordedReject = reject;
 
     dispatch_async(self.sessionQueue, ^{
         [self updateFlashMode];
-        NSString *path = [RNFileSystem generatePathInDirectory:[[RNFileSystem cacheDirectoryPath] stringByAppendingString:@"Camera"] withExtension:@".mov"];
-        NSURL *outputURL = [[NSURL alloc] initFileURLWithPath:path];
-        [self.movieFileOutput startRecordingToOutputFileURL:outputURL recordingDelegate:self];
-        self.videoRecordedResolve = resolve;
-        self.videoRecordedReject = reject;
     });
 }
 
-- (void)stopRecording
+- (void)stopAssetWriter {
+    self.canAppendBuffer = NO;
+    [self.writerInput markAsFinished];
+    [self.videoWriter finishWritingWithCompletionHandler:^{
+        if (self.videoWriter.status != AVAssetWriterStatusFailed) {
+            [self processVideoToAnimation:self.videoWriter.outputURL];
+        } else if (self.videoRecordedReject != nil) {
+            self.videoRecordedReject(@"E_RECORDING_FAILED", @"An error occurred while recording a video.", nil);
+        }
+
+        [self.session removeOutput:self.videoOutput];
+        self.videoOutput = nil;
+        self.writerInput = nil;
+        self.videoWriter = nil;
+    }];
+}
+
+- (void)captureOutput:(AVCaptureOutput *)output didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer fromConnection:(AVCaptureConnection *)connection
+{
+    [self findPrimaryFace:sampleBuffer];
+
+    if (self.canAppendBuffer) {
+        if (self.videoWriter.status != AVAssetWriterStatusWriting) {
+            [self.videoWriter startWriting];
+            [self.videoWriter startSessionAtSourceTime:CMSampleBufferGetPresentationTimeStamp(sampleBuffer)];
+            dispatch_async(dispatch_get_main_queue(), ^{
+                self.timer = [NSTimer scheduledTimerWithTimeInterval:self.maxDuration target:self selector:@selector(stopAssetWriter) userInfo:nil repeats:NO];
+            });
+        }
+        [self.writerInput appendSampleBuffer:sampleBuffer];
+    }
+}
+
+-(void)findPrimaryFace:(CMSampleBufferRef)sampleBuffer {
+    CVPixelBufferRef pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer);
+    CIImage *image = [CIImage imageWithCVPixelBuffer:pixelBuffer];
+    CIImage *orientedImage = [image imageByApplyingCGOrientation:kCGImagePropertyOrientationUpMirrored];
+
+    VNDetectFaceRectanglesRequest *faceDetectionReq = [VNDetectFaceRectanglesRequest new];
```

Comment:
Does a new VNDetectFaceRectanglesRequest need to be created for each new buffer? Totally fine if it does, but just want to make sure we're using memory responsibly. Have you checked for memory leaks with this new code or is everything being garbage collected efficiently?

Reply:
A new VNDetectFaceRectanglesRequest does need to be created for each time we want to search a buffer for a face. Looking at the Profiling tool, memory usage actually never goes above 5% (40% was the CPU) and not seeing any significant memory leaks. So I think we should be ok here. If this becomes an issue, we could reduce the interval at which we check for faces.